Add powerman agent

ofaaland opened the pull request:
Add a fence agent for use with powerman, https://github.com/chaos/powerman.

jnpkrn commented:
Do you really prefer to carry a dedicated and extra maintenance
requiring man page?  It's customary to have the man page automatically
generated based on the metadata, see `fence/agents/lib/fence2man.xsl`.

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,300 @@
+	return outlets
+
+
+def define_new_opts():
```
remove this function, if you don't need new options

ofaaland replied:
retained it because I added --powerman-path

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,300 @@
+		'no_login',
+	]
+
+	raw_args=sys.argv
```
This is something, what you don't want to have in the code. While I agree that it should be fine to see raw input and it can be added into fencing library, your solution did not count with using STDIN as an input method. And stdin is the standard way how to handle arguments in cluster environment.

ofaaland replied:
I removed it here.  I will look at adding to library, with support for both stdin and argv, for debugging.

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,300 @@
+	## Do the delay of the fence device 
+	run_delay(options)
+
+	if options["--action"] in ["off", "reboot"]:
```
I don't understand a reason for this wait

ofaaland replied:
Me either.  This project was started by someone else.  I'll remove it and test.

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,300 @@
+	return
+
+
+def reboot(conn, options):
```
You don't need this function if you don't use --method cycle (unconfirmable reboot). The 'standard' reboot function will work very similar to what you have with some extra features.

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,300 @@
+	logging.debug("get_power_status function:\noptions: %s\n", str(options))
+	pm = PowerMan(options['--ip'], options['--ipport'])
+	# if Pacemaker is checking the status of the Powerman server...
+	if options['--action'] == 'monitor':
```
Monitor action tests if fence device is running what you do test correctly. But 'response' off does not mean that fence device is off but it means that computer is off -> so monitor action will suceed. You have to return something else e.g. 'error' when PM is not running

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,300 @@
+	return True
+
+
+def get_list(conn, options):
```
This function (as referenced from fence_action) should return a lists of machines/plugs that can be turned ON/OFF. You don't have --plug option or any other way so I suspect that you can fence only one machine. Is it right?

ofaaland replied:
powerman can fence a set of machines, e.g. `powerman --off wopr,pipboy[2000,3000]`.  It sounds like we've not implemented this properly.

ofaaland replied:
get_list() does in fact return a dictionary of machines and their power state ("on" or "off") - all the machines that the powerman server knows about.
I thought that was what get_outlet_list function was supposed to do.  Did I get that wrong?

marxsk replied:
ok, so please add 'port' to device_opt array. This will enable --plug used to define which plug/machine/VM should we work with. (Yes, names are inconsistent but we can't fix them right now because of backward compatibility).

get_list - this is right,

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,300 @@
+			return -1
+		logging.debug("pm.on result: %s ret_code: %s\n", result, ret_code)
+
+		## For some devices, the "on" or "off" command can fail silently,
```
That is right point but we have support for this directly in fence library, so you don't have to worry about it.

ofaaland replied:
Do you mean I don't need this block of code:
`
                ## For some devices, the "on" or "off" command can fail silently,
                ## but for query powerman keeps trying until it gets a valid answer;
                ## so we must check after trying to power on/off a device.
                ## Powerman seems to report the old state if asked too quickly, wait a second.

                time.sleep(1)
                queryret = self.query(host)

                if (ret_code == 0 and queryret == "off") or (ret_code < 0 and queryret == "on"):
                        logging.warning("command '%s' returned (%s) but query reported (%s) afterwards\n", cmd, result, queryret)

                if queryret == "on":
                        return 0
                else:
                        return -1
`

and should just return ret_code from the "on" or "off" commands?

marxsk replied:
yes, to be precise we don't care about return codes from set functions at all. Because quite often device really believes that everything is okay but it is not.

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,300 @@
+	"""
+	program_name = "powerman"
+
+	def __init__(self, server_name, port="10101"):
```
magic number 10101 is not required here because you have defined default ipport so either it will be defined by user or default value will be used.

ofaaland replied:
done

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,300 @@
+	elssd8, or whatever values are setup in powerman.conf (***this is key, 
+	because otherwise this code will not work!)
+	"""
+	program_name = "powerman"
```
Please, make the path to program configurable (e.g. like fence_ipmi & ipmitools). We attempt to not have constant across code.

ofaaland replied:
done.

ofaaland commented:
I pushed an update that addresses all the feedback except for the question about get_list(), because I don't yet understand what is required.

ofaaland commented:
The test failed because fence_powerman was run and found that the powerman binary was missing.  Not surprising that you wouldn't have powerman installed.  How do you handle such dependencies?

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,286 @@
+		"shortdesc" : "Path to powerman binary",
+		"default" : "/usr/bin/powerman",
+		"order": 200
+	}
```
Please chnage default value to "@POWERMAN_PATH@" and add line to configure.in (line ~185 at root dir) and to make/fencebuild.mk 

This way the default can be distribution dependant.

marxsk commented on fence/agents/powerman/fence_powerman.py:
```diff
@@ -0,0 +1,286 @@
+	if not is_executable(options["--powerman-path"]):
+		fail_usage("Powerman not found or not executable at path " + options["--powerman-path"])
+
+	if options["--action"] in ["off"]:
```
This can be defined by setting --power_wait option to different value. 

Look at https://fedorahosted.org/cluster/wiki/FenceTiming

ofaaland commented:
Pushed another update that addressed the comments from pass 2:
* default path for powerman set in configure
* remove redundant query within PowerMan.on() and off() used by set_power_fn
* use run_delay() before testing whether powerman binary exists and is executable

marxsk commented:
Thanks a lot for new fence agent, it will be part of next upstream release.
